### ISSUE TITLE
Video support

### ISSUE THREAD

**Issue by viliusle** [rqrauhvmra/Tobi#3]
Video (or at least youtube) support would be nice to have.

**Comment by rqrauhvmra** [comment]
Hi!

Yes. The possibility of adding HTML elements would also be great.

**Comment by viliusle** [comment]
i added youtube support - https://github.com/viliusle/Tobi (there is demo, but minified code was not updated)

p.s. why you never use ";" in your code? Other people will get 999+ warnings ... Same about no tabs, and tab size, but that is off-topic.

**Comment by midzer** [comment]
Great :+1: 
Do `<video>` elements work, too?

**Comment by viliusle** [comment]
No, only youtube, for now. in 90% cases videos are from youtube :)

Todo for future:

- optimize code to make it easy to add new types (see list items below)
- support <video>
- support HTML
- support iframes with external pages (?)

**Comment by viliusle** [comment]
Added support for youtube videos, HTML5 videos, iframes, inline HTML.
https://github.com/viliusle/Tobi

**Need help with testing!** Download code below, open demo.html and test. Once everybody are happy, me or author should merge.

**Comment by midzer** [comment]
`demo.html` is doing smoothly for me. Awesome work, @viliusle !

Maybe in upcoming days I can have a closer look at details in `tobi.js`. Looks like alot of changes :)

**Comment by rqrauhvmra** [comment]
Hi @viliusle and @midzer 

I'm on vacation this week and will take a look at it next week. Thanks for your support.

**Comment by rqrauhvmra** [comment]
Hi @viliusle 

I've merged your fork into [the dev branch](https://github.com/rqrauhvmra/Tobi/tree/dev) but I'll have to test it. Regarding your question why I never use semicolons, I use [JavaScript Standard Style](https://standardjs.com/#standardjs--the-rules). I'll declare it in the package.json and in the README.md soon.

Thanks again for your support.

**Comment by midzer** [comment]
Last days I've played around with latest changes and got HTML5 videos working in production https://feuerwehr-eisolzried.de/media/

Only things I had to change:
* https://github.com/rqrauhvmra/Tobi/blob/ecd78fe212f0f3a04a78b9b91122f4672100adbf/js/tobi.js#L344 -> change to `var target = element.querySelector(targetSelector)` _and_ move `video` inside `a`:
```
<a href="#" data-type="html" data-target="#selectorVideo" class="lightbox">
    Video (HTML5)
    <div style="display:none;" id="selectorVideo">
         <video controls src="https://interactive-examples.mdn.mozilla.net/media/examples/stream_of_water.webm">
        </video>
    </div>
</a>
```
Note, that I use dynamic adding of elements via `add()`. Maybe thats why my changes were necessary.

I wonder whether it is possible to drop the `<div style="display:none;" id="selectorVideo">` container alltogether and simply make this style and id an requirement for `video` or custom HTML elements.

* another change is definetely discussable:
in `onLoad` for HTML elements at https://github.com/rqrauhvmra/Tobi/blob/ecd78fe212f0f3a04a78b9b91122f4672100adbf/js/tobi.js#L365 I've inserted:
```
var video = container.querySelector('video');
if (video != undefined){
    video.load();
    video.play();
}
```
to make `video` elements load and play directly when lightbox kicks in. This comes especially hand for `video` elements with `metadata="none"` to save bandwidth.

Furthermore I've detected minor issues in mobile view for vertical videos that extend viewport. But maybe I've got to play around more with my own CSS classes.

So much for the input and have a nice week
midzer

**Comment by rqrauhvmra** [comment]
Hi @midzer 

Thanks for your feedback.

I've problems with your first change, referring to the demo.html. If I change `target`, I also must move the inline html div inside `a` but the `a`-Tag inside this div breaks the lighbox ([Codepen demo](https://codepen.io/rqrauhvmra/pen/gjEeaV)).

**Comment by midzer** [comment]
Okay, after digging some more into issue 1, I've found an internal problem. So nothing to do with @viliusle  new changes on Tobi: There is no need at all to place `<div style="display:none;" id="selector">` right inside `a` for dynamic content. Sorry :|

On the other hand, we might talk about changing to `var target = document.getElementById(targetSelector)` while using `data-target="selectorVideo"` for minor byte savings and performance reasons https://jsperf.com/getelementbyid-vs-queryselector

**Comment by viliusle** [comment]
getElementById is more strict and accept only ids, while querySelector is more flexible (for price of being just a little slower, like 0.000x). So question is do we want to allow only ids or general selectors.

**Comment by midzer** [comment]
In previously linked test `getElementById()` is 3x as fast as `querySelector` on my machine. Yeah, the latter is way more flexible to allow selection via `class` or other attributes for example, but slower functions add up each time for large projects. I think deeply investigating their best usage is important.

The question is if there's really a use case to allow more flexible selection method for an individual `data-html` target. I'm having a hard time thinking of multiple `selectorTarget`s to be triggered and shown in lightbox.

**Comment by viliusle** [comment]
You last argument is strong, i agree we can switch to getElementById.
p.s. what i wanted to say that both are fast enough and difference is too small to spend lots of time thinking about it :)

Another missing feature could be missing resize event for youtube on phone. (if user rotate phone, is big youtube videos still ok - need testing).

p.s. i can not help with code anymore, only with suggestions.

**Comment by rqrauhvmra** [comment]
Hi

I've published the new version without the YouTube support because I have some issues.

Your video example worked @viliusle but some YouTube Links which I get from the share function on YouTube don't work.

For example:
https://www.youtube.com/embed/AI_mzJXUFpk or
https://www.youtube.com/embed/VYOjWnS4cMY

I'll integrate the YouTube Iframe-API soon.

**Comment by viliusle** [comment]
These videos work only on youtube domain, dont know why, but even using official embed code below - video dont work o_O

`<iframe width="560" height="315" src="https://www.youtube.com/embed/VYOjWnS4cMY" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>`

p.s. would you include demo.html somwhere? since its good place to test all object types when developing.

**Comment by rqrauhvmra** [comment]
Yes, I include a demo.html next week. I wanted to make this a little more detailed.

**Comment by midzer** [comment]
Some progress on `video` element support in https://github.com/midzer/Tobi/tree/autoVideoPlayback
* added userOption `autoplayVideo: false`
* remember `currentTime` of video on slide/close (for potential revisit in order to continue from last timestamp)
* stop fetching data on unfinished video load if user slides or closes Tobi
* temporarily comment out `overflow-y: auto` for `html` elements to make viewport overflowing, large resolution videos proper scale
* minor CSS enhancements
* some refactoring

Live Demo https://feuerwehr-eisolzried.de/media/ (Tab "Videos")

Have a nice week
midzer

**Comment by rqrauhvmra** [comment]
Hi @midzer 

Looks great, [I have added two comments](https://github.com/midzer/Tobi/commit/632ee74c319fc800cd895ec3d27f69d341cedb37).

**Comment by rqrauhvmra** [comment]
Hi @midzer 

I use HTML5 Videos with `source` elements, not with the `src` attribute.

```html
<video controls>
  <source src="https://staging.coverr.co/s3/mp4/Flam.mp4" type="video/mp4">
  Your browser does not support the video tag.
</video>
```

In this case your `onCleanup()` failed and the video is gone.

**Comment by midzer** [comment]
@rqrauhvmra Howdy,
yeah, I did not respect videos with `source` yet :[
Let me create a patch for this.

**Comment by rqrauhvmra** [comment]
Good morning @midzer 

I changed your patch a bit. In Firefox 62 it sets all `src` attributes to `data-src` when I close the lightbox and only the video plays completely. When I open it again, the attributes go back to `src` but the video can not be loaded.

In Chrome 68.0.3440.106 it does not work at all.

I haven't found any solution.

**Comment by midzer** [comment]
Awesome, thanks for your testing + fix!

**Comment by rqrauhvmra** [comment]
Hi @midzer 

Concerning your function to save bandwidth with videos, I've tested a little bit (Firefox 62). (Users should integrated videos with `preload="none"` and `poster="link/to/preview.png"`)

My [current code](https://github.com/rqrauhvmra/Tobi/blob/dev/js/tobi.js#L332) looks like this:

```js
onCleanup: function (container) {
  var video = container.querySelector('video')

  console.log('video.readyState: ' + video.readyState)
  console.log('video.currentTime: ' + video.currentTime)
  console.log('video.duration: ' + video.duration)

  if (video) {
    if (video.readyState > 0 && video.readyState < 3 && video.duration !== video.currentTime) {
      // Some data has been loaded but not the whole package.
      // In order to save bandwidth, stop downloading
      // as soon as possible.
      // According to https://developer.mozilla.org/en-US/docs/Web/Apps/Fundamentals/Audio_and_video_delivery#Stopping_the_download_of_media
      // this can be achieved by:
      // 1. backup src
      // 2. remove src
      // 3. call load()
      var clone = video.cloneNode(true)

      setVideoSources(video, 'src', 'data-src')
      video.load()

      video.parentNode.removeChild(video)

      container.appendChild(clone)
    }
  }
}
```

- `video.duration !== video.currentTime` because if the video was fully loaded and complete played, `video.readyState` is `2` (that's crazy)
- If I don't append the video new, it don't work in Firefox. The video can't be loaded.

**Comment by viliusle** [comment]
Please find some time to include youtube support.

**Comment by rqrauhvmra** [comment]
Hi @viliusle 

I do my best and it will soon be ready.

**Comment by rqrauhvmra** [comment]
Hello,

I'm currently [testing the support for YouTube](https://codepen.io/rqrauhvmra/pen/ReMgKY). Things that do not work yet:

- `host: 'https://www.youtube-nocookie.com'` for GDPR / DSGVO

I still implement a few parameters (controls, height, width, showinfo...). I would let "Responsive" be done by an external library.

**Comment by rqrauhvmra** [comment]
So,

looks like I fixed the problem :clap:. Please thoroughly check.

**Comment by midzer** [comment]
For a quick test via demo page linked above YT video works fine for me.

@viliusle what do you think about it (as you requested this feature)?

**Comment by viliusle** [comment]
Is this correct test url? i see bug: open youtube link, click play, click outside, so popup window will disappear, but actual video will still play in background, and you will hear sound, that is annoying.

**Comment by rqrauhvmra** [comment]
Hi @viliusle 

Yes, this is the correct url. You found a bug. I'll fix it tomorrow.

**Comment by rqrauhvmra** [comment]
Good morning @viliusle 

I fixed the issue.

**Comment by rqrauhvmra** [comment]
## Hello

I'll be changing and improving a few things at the beginning next week and releasing version 1.8.0. Thank you for your support so far.

Have a nice weekend.
